File: src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Statements/ReturnKeywordRecommender.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis.Completion.Providers
Imports Microsoft.CodeAnalysis.VisualBasic.Extensions.ContextQuery

Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.KeywordRecommenders.Statements
    ''' <summary>
    ''' Recommends the "Return" keyword at the start of a statement
    ''' </summary>
    Friend Class ReturnKeywordRecommender
        Inherits AbstractKeywordRecommender

        Private Shared ReadOnly s_keywords As ImmutableArray(Of RecommendedKeyword) =
             ImmutableArray.Create(New RecommendedKeyword("Return", VBFeaturesResources.Returns_execution_to_the_code_that_called_the_Function_Sub_Get_Set_or_Operator_procedure_Return_or_Return_expression))

        Protected Overrides Function RecommendKeywords(context As VisualBasicSyntaxContext, cancellationToken As CancellationToken) As ImmutableArray(Of RecommendedKeyword)
            Return If(context.IsStatementContext AndAlso Not context.IsInStatementBlockOfKind(SyntaxKind.FinallyBlock),
                s_keywords,
                ImmutableArray(Of RecommendedKeyword).Empty)
        End Function
    End Class
End Namespace

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/ReturnKeywordRecommender.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Threading;
using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.CSharp.Utilities;
using Microsoft.CodeAnalysis.Shared.Extensions;

namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders;

internal class ReturnKeywordRecommender : AbstractSyntacticSingleKeywordRecommender
{
    public ReturnKeywordRecommender()
        : base(SyntaxKind.ReturnKeyword)
    {
    }

    protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
    {
        return
            context.IsStatementContext ||
            context.IsRegularTopLevelStatementsContext() ||
            context.TargetToken.IsAfterYieldKeyword() ||
            IsAttributeContext(context, cancellationToken);
    }

    private static bool IsAttributeContext(CSharpSyntaxContext context, CancellationToken cancellationToken)
    {
        return
            context.IsMemberAttributeContext(SyntaxKindSet.ClassInterfaceStructRecordTypeDeclarations, cancellationToken) ||
            (context.SyntaxTree.IsScript() && context.IsTypeAttributeContext(cancellationToken)) ||
            context.IsStatementAttributeContext() ||
            IsAccessorAttributeContext();

        bool IsAccessorAttributeContext()
        {
            var token = context.TargetToken;
            return token.Kind() == SyntaxKind.OpenBracketToken &&
                token.Parent is AttributeListSyntax &&
                token.Parent.Parent is AccessorDeclarationSyntax;
        }
    }
}
